pyomo/contrib/mindtpy/single_tree.py

```diff
@@ -0,0 +1,257 @@
+from math import fabs
+from pyomo.repn import generate_standard_repn
+import logging
+from pyomo.common.dependencies import attempt_import
+
+cplex, cplex_available = attempt_import('cplex')
+if cplex_available:
+    from cplex.callbacks import LazyConstraintCallback
+else:
+    logging.warning(
+        "Cplex python API is not found. Therefore, lp-nlp is not supported")
+    # Other solvers (e.g. Gurobi) are not supported yet
+
+
+class LazyOACallback_cplex(LazyConstraintCallback):
+    """Inherent class in Cplex to call Lazy callback."""
+
+    def copy_lazy_var_list_values(self, opt, from_list, to_list, config,
+                                  skip_stale=False, skip_fixed=True,
+                                  ignore_integrality=False):
+        """Copy variable values from one list to another.
+
+        Rounds to Binary/Integer if neccessary
+        Sets to zero for NonNegativeReals if neccessary
+        """
+        for v_from, v_to in zip(from_list, to_list):
+            if skip_stale and v_from.stale:
+                continue  # Skip stale variable values.
+            if skip_fixed and v_to.is_fixed():
+                continue  # Skip fixed variables.
+            try:
+                v_to.set_value(self.get_values(
+                    opt._pyomo_var_to_solver_var_map[v_from]))
+                if skip_stale:
+                    v_to.stale = False
+            except ValueError as err:
+                err_msg = getattr(err, 'message', str(err))
+                # get the value of current feasible solution
+                # self.get_value() is an inherent function from Cplex
+                var_val = self.get_values(
+                    opt._pyomo_var_to_solver_var_map[v_from])
+                rounded_val = int(round(var_val))
+                # Check to see if this is just a tolerance issue
+                if ignore_integrality \
+                    and ('is not in domain Binary' in err_msg
+                         or 'is not in domain Integers' in err_msg):
+                    v_to.value = self.get_values(
+                        opt._pyomo_var_to_solver_var_map[v_from])
+                elif 'is not in domain Binary' in err_msg and (
+                        fabs(var_val - 1) <= config.integer_tolerance or
+                        fabs(var_val) <= config.integer_tolerance):
+                    v_to.set_value(rounded_val)
+                # TODO What about PositiveIntegers etc?
+                elif 'is not in domain Integers' in err_msg and (
+                        fabs(var_val - rounded_val) <= config.integer_tolerance):
+                    v_to.set_value(rounded_val)
+                # Value is zero, but shows up as slightly less than zero.
+                elif 'is not in domain NonNegativeReals' in err_msg and (
+                        fabs(var_val) <= config.zero_tolerance):
+                    v_to.set_value(0)
+                else:
+                    raise
```

Comment:
I think that you could make this more robust by doing something like:
```python
try:
    v_val = self.get_values(opt._pyomo_var_to_solver_var_map[v_from])
    v_to.set_value(v_val)
    if skip_stale:
        v_to.stale = False
except ValueError:
    # Snap the value to the bounds
    if v_to.lb is not None and v_val < v_to.lb and v_to.lb - v_val <= config.zero_tolerance:
        v_to.set_value(v_to.lb)
    elif v_to.ub is not None and v_val > v_to.ub and v_val - v_to.ub <= config.zero_tolerance:
        v_to.set_value(v_to.ub)
    #... or the nearest integer
    elif v_to.is_integer():
        rounded_val = int(round(v_val))
        if (ignore_integrality or fabs(v_val - rounded_val) <= config.integer_tolerance) \
                and rounded_val in v_to.domain:
            v_to.set_value(rounded_val)
    else:
        raise
```

Reply:
Thanks for the advice and the code.